src/Data/LLVM/BitCode/BitString.hs

```diff
@@ -100,31 +111,36 @@ emptyBitString = BitString (NumBits 0) 0
 -- BitString.
 
 joinBitString :: BitString -> BitString -> BitString
-joinBitString (BitString (Bits' (I# szA#)) (I# a#))
-              (BitString (Bits' (I# szB#)) (I# b#)) =
+joinBitString (BitString (Bits' (I# szA#)) (W# a#))
+              (BitString (Bits' (I# szB#)) (W# b#)) =
   BitString { bsLength = NumBits (I# (szA# +# szB#))
-            , bsData = I# (a# `orI#` (b# `uncheckedIShiftL#` szA#))
+            , bsData = W# (a# `or#` (b# `uncheckedShiftL#` szA#))
             }
 
+bitMask :: NumBits -> Word#
+bitMask (Bits' (I# len#)) =
+  ((int2Word# 1#) `uncheckedShiftL#` len#) `minusWord#` (int2Word# 1#)
+
 
 -- | Given a number of bits to take, and an @Integer@, create a @BitString@.
 
-toBitString :: NumBits -> Int -> BitString
-toBitString len@(Bits' (I# len#)) (I# val#) =
-  let !mask# = (1# `uncheckedIShiftL#` len#) -# 1#
-  in BitString len (I# (val# `andI#` mask#))
+toBitString :: NumBits -> Word -> BitString
+toBitString len (W# val#) = BitString len (W# (val# `and#` (bitMask len)))
 
 
 -- | Extract the referenced Integer value from a BitString
 
-bitStringValue :: BitString -> Int
+bitStringValue :: BitString -> Word
 bitStringValue = bsData
 
 
 -- | Extract a target (Num) value of the desired type from a BitString (using
 -- fromInteger to perform the target type conversion).
 
 fromBitString :: (Num a, Bits a) => BitString -> a
+#ifdef QUICK
+fromBitString (BitString _ i) = x
+#else
```

Comment:
I'm not a fan of sprinkling CPP everywhere throughout the library like this, since (1) it makes adds a lot of lexical noise, and (2) it makes it more difficult to remember to keep the two versions of every function in sync, since you have to remember to build with both `-fquick` and `-f-quick` unless you want to risk bitrot.

Rather this going with this approach, why not do something like this instead?

```hs
fromBitString :: (Num a, Bits a) => BitString -> a
fromBitString (BitString l i)
  | quickMode = x
  | otherwise =
    case bitSizeMaybe x of
      ...
```

Where:

```hs
quickMode :: Bool
#if defined(QUICK)
quickMode = True
#else
quickMode = False
#endif
```

This way, we can continue to typecheck both versions of each function, decreasing the chances of bitrot. Since `quickMode` will statically be known to be `True` or `False`, the optimizer will be able to make swift work of it.